js/render/draw_line.js

```diff
@@ -119,45 +36,122 @@ module.exports = function drawLine(painter, source, layer, coords) {
 
-        } else {
-            gl.uniform1f(program.u_ratio, ratio);
-        }
+        painter.enableTileClippingMask(coord);
 
         for (var i = 0; i < bufferGroups.length; i++) {
             var group = bufferGroups[i];
-            group.vaos[layer.id].bind(gl, program, group.layoutVertexBuffer, group.elementBuffer);
+            group.vaos[layer.id].bind(gl, program, group.layoutVertexBuffer, group.elementBuffer, group.paintVertexBuffers[layer.id]);
             gl.drawElements(gl.TRIANGLES, group.elementBuffer.length * 3, gl.UNSIGNED_SHORT, 0);
         }
     }
-
 };
+
+
+function bindLineSDFPatternProgram(painter, source, layer, tile, programOptions) {
+    var gl = painter.gl;
+
+    var program = painter.useProgram(
+        'linesdfpattern',
+        programOptions.defines,
+        programOptions.vertexPragmas,
+        programOptions.fragmentPragmas
+    );
+
+    var dasharray = layer.paint['line-dasharray'];
+    var posA = painter.lineAtlas.getDash(dasharray.from, layer.layout['line-cap'] === 'round');
+    var posB = painter.lineAtlas.getDash(dasharray.to, layer.layout['line-cap'] === 'round');
+    var widthA = posA.width * dasharray.fromScale;
+    var widthB = posB.width * dasharray.toScale;
+    painter.setUniforms({
+        'u_tex_y_a': posA.y,
+        'u_tex_y_b': posB.y,
+        'u_mix': dasharray.t,
+        'u_patternscale_a': [1 / pixelsToTileUnits(tile, widthA, painter.transform.tileZoom), -posA.height / 2],
+        'u_patternscale_b': [1 / pixelsToTileUnits(tile, widthB, painter.transform.tileZoom), -posB.height / 2],
+        'u_sdfgamma': painter.lineAtlas.width / (Math.min(widthA, widthB) * 256 * browser.devicePixelRatio) / 2
+    });
+    gl.uniform1i(program.u_image, 0);
+    gl.activeTexture(gl.TEXTURE0);
+    painter.lineAtlas.bind(gl);
+
+    return program;
+}
+
+function bindLinePatternProgram(painter, source, layer, tile, programOptions) {
+    var gl = painter.gl;
+
+    var program = painter.useProgram(
+        'linepattern',
+        programOptions.defines,
+        programOptions.vertexPragmas,
+        programOptions.fragmentPragmas
+    );
+
+    var image = layer.paint['line-pattern'];
+    var imagePosA = painter.spriteAtlas.getPosition(image.from, true);
+    var imagePosB = painter.spriteAtlas.getPosition(image.to, true);
+    if (!imagePosA || !imagePosB) return;
+    painter.setUniforms({
+        'u_pattern_tl_a': imagePosA.tl,
+        'u_pattern_br_a': imagePosA.br,
+        'u_pattern_tl_b': imagePosB.tl,
+        'u_pattern_br_b': imagePosB.br,
+        'u_fade': image.t,
+        'u_pattern_size_a': [
+            pixelsToTileUnits(tile, imagePosA.size[0] * image.fromScale, painter.transform.tileZoom),
+            imagePosB.size[1]
+        ],
+        'u_pattern_size_b': [
+            pixelsToTileUnits(tile, imagePosB.size[0] * image.toScale, painter.transform.tileZoom),
+            imagePosB.size[1]
+        ]
+    });
```

Comment:
Since this is in a hot path, it might be worth trying to rewrite that as multiple `painter.setUniform(name, value)` statements to avoid the `for in` loop and temporary objects. I would even do 4 different methods (`setUniform2`, etc.) to avoid temporary arrays or the `arguments` keyword.